src/validate.js

```diff
@@ -159,6 +160,308 @@ function transformAjvErrors(errors = []) {
+ * When one is found a new property is added to the `_target` with the path to
+ * this `__errors` as the key and the value of `__errors`.
+ *
+ * @example
+ * // sample input
+ * {
+ *     foo: {
+ *         bar: {
+ *             __errors: [...__errorsOfThisProp],
+ *             2: {
+ *                 __errors: [...__errorsOfThisProp]
+ *             }
+ *         }
+ *     }
+ * }
+ * // sample output
+ * // different cases possible of an error.properties; here are some EXAMPLES
+ * {
+ *     ".foo": [* errors like: a -> is-required; is-not-a-string; is-not-of-type *],
+ *     ".bar[n]": [* errors like: b -> is-not-an-array *],
+ *     ".bar[n].qux": [* errors like: (a) *],
+ *     ".bar[n].qux[4]": [* errors like: (b) *],
+ *     ".bar[n].qux[4].wibble": [* errors like: (a) *],
+ *     "": [* errors like: does-not-match-oneof-schema *]
+ * }
+ *
+ * @param {object} _target - the object to add the key to.
+ * @param {object[]} _errors - the errors of ajv
+ * @param {object} _errorSchema - the error schema constructed by the `toErrorSchema` function
+ * @param {string} _currentPath - the path of the current nested object
+ */
+function flattenErrorSchema(_target, _errors, _errorSchema, _currentPath = "") {
+  Object.entries(_errorSchema).forEach(([schemaKey, schemaValue]) => {
+    if (Array.isArray(schemaValue)) {
+      // we found the __errors array
+      // find the error in _errors for this path
+      const pathWithoutErrorsArray = _currentPath
+        .split(".")
+        .filter(x => x !== "__errors");
+      _target[_currentPath] = _errors.filter(
+        err =>
+          err.property ===
+          `.${pathWithoutErrorsArray.reduce(
+            (acc, curVal) =>
+              `${acc}${
+                err.property && err.property.includes(`[${curVal}]`)
+                  ? `[${curVal}]`
+                  : `.${curVal}`
+              }`
+          )}`
+      );
+    } else {
+      // we did not find the __errors array, yet. Let's go deeper
+      let nextLevelPath;
+      if (
+        _errors.some(
+          err =>
+            "property" in err &&
+            err.property.includes(`${_currentPath}[${schemaKey}]`)
+        )
+      ) {
+        // array!
+        nextLevelPath = `${_currentPath}[${schemaKey}]`;
+      } else {
+        nextLevelPath = [_currentPath, schemaKey].filter(x => !!x).join(".");
+      }
+      flattenErrorSchema(_target, _errors, schemaValue, nextLevelPath);
+    }
+  });
+}
+
+/**
+ * sorting based on flattened _errorSchema
+ *
+ * @param _uiSchema
+ * @param _errors
+ * @param _errorSchema
+ * @return {Array}
+ */
+function orderErrorsByUiSchema(_uiSchema, _errors, _errorSchema) {
+  if (!Array.isArray(_errors) || _errors.length === 0) {
+    return [];
+  }
+  console.log(`in`, _errorSchema);
+  const flattenedErrorSchema = {};
+  flattenErrorSchema(flattenedErrorSchema, _errors, _errorSchema);
+  console.log(`out`, flattenedErrorSchema);
```

Comment:
My added `Question 3` can be checked by adding the `.only` to the suite and looking at these loggings.